source/mainnet/smart-contracts/tutorials/sponsoredTransactions/sponsoredTransactionsSmartContract.rst

```diff
@@ -0,0 +1,247 @@
+that allows a special trusted role/account to register the public keys in the smart contract. The example smart contract implements
+such a registry. The logic of the code is as such that once an account has a public key registered, the mapping between the public key and the account can not be
+updated anymore in the `public_key_registry` as seen below:
+
+.. code-block:: rust
+
+    // Register the public key.
+    let old_public_key = host.state_mut().public_key_registry.insert(param.account, (param.public_key, 0));
+
+    // Return an error if the owner tries to update/re-write a new public key for an already registered account.
+    ensure!(old_public_key.is_none(), CustomContractError::AccountAlreadyRegistered.into());
+
+The registry in the smart contracts links an account to a public key and its next nonce.
+The nonce is increased sequentially every time a signed message (corresponding to the account) is successfully executed in the ``permit`` function.
+The nonce is used to prevent replay attacks of the signed message because only a message signed with the next nonce in sequence can be executed as seen below:
+
+.. code-block:: rust
+
+    // Check the nonce to prevent replay attacks.
+    ensure_eq!(message.nonce, nonce, CustomContractError::NonceMismatch.into());
+
+Testing the signature verification
+==================================
+
+In this section, you will learn how to create an Ed25519 signature
+that you can use for your test cases to confirm that the smart contract signature verification logic works as expected.
+
+.. note::
+    The commands in this section have to be executed from the `./examples/cis3-nft-sponsored-txs` folder where the ``Cargo.toml`` file is defined for this project.
+
+Run the test cases to confirm that they pass before you start modifying the code:
+
+.. code-block:: console
+
+    $cargo test
+    or
+    $cargo concordium test
+
+.. image:: ./images/testCases.png
+   :alt: Running the test cases
+   :align: center
+
+You should see no errors and all the test cases are green indicating that they passed successfully.
+
+.. note::
+
+   Comprehensive instructions on how to download and set up  ``cargo concordium`` can be found in :ref:`Setup the development environment<setup-env>`.
+
+The next few commands explain how you can display the ``message_hash`` that you would need to sign.
+Since hashes are commonly displayed as hex numbers, add the ``hex`` package to your project to convert and display such values easily.
+
+.. code-block:: rust
+
+    cargo add hex
+
+You can check if the ``hex`` package was added in the ``Cargo.toml`` file under ``[dependencies]``.
+
+.. image:: ./images/CargoToml.png
+   :alt: Cargo.toml file with hex crate
+   :align: center
+
+Add the following to the smart contract ``permit`` function after the variable ``message_hash`` is defined.
+
+.. code-block:: rust
+
+    println!("{:?}", hex::encode(message_hash.to_vec()));
```

Comment:
I appreciate the cleverness of this solution, but would it not be better to add a "view" function that would return the message hash given parameters? Then that could be run either with concordium-client or locally with cargo-concordium run?